packages/protocol/contracts/common/FeeHandler.sol

```diff
@@ -454,28 +672,54 @@ contract FeeHandler is
     emit SoldAndBurnedToken(tokenAddress, balanceToBurn);
   }
 
+  function _calculateDistributeAmounts(
+    FixidityLib.Fraction memory thisTokenFraction,
+    FixidityLib.Fraction memory totalFractionOfOtherBeneficiariesAndCarbonFixidity,
+    uint256 toDistribute
+  ) private returns (uint256) {
+    FixidityLib.Fraction memory proportionOfThisToken = thisTokenFraction.divide(
+      totalFractionOfOtherBeneficiariesAndCarbonFixidity
+    );
+
+    FixidityLib.Fraction memory toDistributeFraction = FixidityLib.newFixed(toDistribute).multiply(
+      proportionOfThisToken
+    );
+
+    return toDistributeFraction.fromFixed();
+  }
+
   function _distribute(address tokenAddress) private onlyWhenNotFrozen nonReentrant {
-    require(feeBeneficiary != address(0), "Can't distribute to the zero address");
+    require(carbonFeeBeneficiary != address(0), "Can't distribute to the zero address");
     IERC20 token = IERC20(tokenAddress);
     uint256 tokenBalance = token.balanceOf(address(this));
 
     TokenState storage tokenState = tokenStates[tokenAddress];
-    require(
-      tokenState.handler != address(0) ||
-        tokenAddress == registry.getAddressForOrDie(CELO_TOKEN_REGISTRY_ID),
-      "Handler has to be set to sell token"
+
+    FixidityLib.Fraction
+      memory totalFractionOfOtherBeneficiariesAndCarbonFixidity = getTotalFractionOfOtherBeneficiariesAndCarbonFixidity();
+
+    uint256 carbonFundAmount = _calculateDistributeAmounts(
+      getCarbonFractionFixidity(),
+      totalFractionOfOtherBeneficiariesAndCarbonFixidity,
+      tokenState.toDistribute
     );
 
-    // safty check to avoid a revert due balance
-    uint256 balanceToDistribute = Math.min(tokenBalance, tokenState.toDistribute);
+    _executePayment(tokenAddress, tokenState, carbonFeeBeneficiary, carbonFundAmount);
 
-    if (balanceToDistribute == 0) {
-      // don't distribute with zero balance
-      return;
+    for (uint256 i = 0; i < EnumerableSet.length(otherBeneficiariesAddresses); i++) {
+      address beneficiary = otherBeneficiariesAddresses.get(i);
+      Beneficiary storage otherBeneficiary = otherBeneficiaries[beneficiary];
+
+      uint256 amount = _calculateDistributeAmounts(
```

Comment:
Originally we had here safety measure: uint256 balanceToDistribute = Math.min(tokenBalance, tokenState.toDistribute); to be sure not to distribute more than we actually have - should we add it back ?

Reply:
that could break internal accounting now, I'd prefer it to revert